crates/ruff/src/rules/pyflakes/rules/f_string_missing_placeholders.rs

```diff
@@ -47,45 +47,72 @@ impl AlwaysAutofixableViolation for FStringMissingPlaceholders {
-    expr: &'a Expr,
+/// Return an iterator containing a two-element tuple for each f-string in the
+/// given [`ExprFString`] expression.
+///
+/// The first element of the tuple is the f-string prefix range, and the second
+/// element is the entire f-string range. It returns an iterator because of the
+/// possibility of multiple f-strings implicitly concatenated together.
+///
+/// For example,
+///
+/// ```python
+///   f"first" rf"second"
+/// # ^         ^            (prefix range)
+/// # ^^^^^^^^ ^^^^^^^^^^    (token range)
+/// ```
+///
+/// would return `[(0..1, 0..8), (10..11, 9..19)]`.
+///
+/// This function assumes that the given f-string expression is without any
+/// placeholder expressions.
+///
+/// [`ExprFString`]: `ruff_python_ast::ExprFString`
+fn fstring_prefix_and_tok_range<'a>(
+    fstring: &'a ast::ExprFString,
     locator: &'a Locator,
     source_type: PySourceType,
 ) -> impl Iterator<Item = (TextRange, TextRange)> + 'a {
-    let contents = locator.slice(expr);
-    lexer::lex_starts_at(contents, source_type.as_mode(), expr.start())
+    let contents = locator.slice(fstring);
+    let mut current_f_string_start = fstring.start();
+    lexer::lex_starts_at(contents, source_type.as_mode(), fstring.start())
         .flatten()
-        .filter_map(|(tok, range)| match tok {
-            Tok::String {
-                kind: StringKind::FString | StringKind::RawFString,
-                ..
-            } => {
-                let first_char = locator.slice(TextRange::at(range.start(), TextSize::from(1)));
+        .filter_map(move |(tok, range)| match tok {
+            Tok::FStringStart => {
+                current_f_string_start = range.start();
+                None
+            }
+            Tok::FStringEnd => {
+                let first_char =
+                    locator.slice(TextRange::at(current_f_string_start, TextSize::from(1)));
                 // f"..."  => f_position = 0
                 // fr"..." => f_position = 0
                 // rf"..." => f_position = 1
                 let f_position = u32::from(!(first_char == "f" || first_char == "F"));
                 Some((
                     TextRange::at(
-                        range.start() + TextSize::from(f_position),
+                        current_f_string_start + TextSize::from(f_position),
                         TextSize::from(1),
                     ),
-                    range,
+                    TextRange::new(current_f_string_start, range.end()),
                 ))
             }
             _ => None,
         })
 }
 
 /// F541
-pub(crate) fn f_string_missing_placeholders(expr: &Expr, values: &[Expr], checker: &mut Checker) {
+pub(crate) fn f_string_missing_placeholders(
+    fstring: &ast::ExprFString,
+    values: &[Expr],
+    checker: &mut Checker,
+) {
     if !values
         .iter()
         .any(|value| matches!(value, Expr::FormattedValue(_)))
     {
         for (prefix_range, tok_range) in
```

Comment:
>  It's also unclear to me where the filtering out of f-strings without placeholders happens.

@MichaReiser

Reply:
Thanks. The issue I see is that `find_useless_f_strings` returns all fstrings but with potentially incorrect ranges. So naming that function `find_useless_f_strings` is misleading in my view because not all fstrings returned by that function are useless.

Reply:
Yes, I totally agree. I've renamed it to `fstring_prefix_and_tok_range` but the main reason it returns an iterator is because the f-string can contain multiple implicitly concatenated (f-)strings.

> because not all fstrings returned by that function are useless.

Can you expand on this? For example, in the below code snippet it would return (prefix range, token range) for the "first" and the "last" string and not for the "normal" string. Both f-strings are useless in the sense that there are no placeholders used and so the `f` prefix is useless. I don't think it ever returns incorrect ranges.

```python
  f"first" "normal" rf"second"
# ^                  ^             (prefix range)
# ^^^^^^^^          ^^^^^^^^^^     (token range)

# (0..1, 0..8)      - `f"first"`
# (19..20, 18..28)  - `rf"second"`
```